csharp/ql/src/semmle/code/csharp/dataflow/FlowSummary.qll

```diff
@@ -0,0 +1,145 @@
+/**
+ * Provides classes and predicates for definining flow summaries.
+ */
+
+import csharp
+private import internal.FlowSummaryImpl as Impl
+private import internal.FlowSummarySpecific::Private
+private import internal.DataFlowPublic as DataFlowPublic
+// import all instances below
+private import semmle.code.csharp.dataflow.LibraryTypeDataFlow
+
+class SummarizableCallable = Impl::SummarizableCallable;
+
+/** An unbound method. */
+class SummarizableMethod extends SummarizableCallable, Method { }
+
+class ContentList = Impl::ContentList;
+
+/** Provides predicates for constructing content lists. */
+module ContentList {
+  import Impl::ContentList
+
+  /** Gets the singleton "element content" content list. */
+  ContentList element() { result = singleton(any(DataFlowPublic::ElementContent c)) }
+
+  /** Gets a singleton property content list. */
+  ContentList property(Property p) {
+    result =
+      singleton(any(DataFlowPublic::PropertyContent c | c.getProperty() = p.getSourceDeclaration()))
+  }
+
+  /** Gets a singleton field content list. */
+  ContentList field(Field f) {
+    result =
+      singleton(any(DataFlowPublic::FieldContent c | c.getField() = f.getSourceDeclaration()))
+  }
+}
+
+class SummaryInput = Impl::SummaryInput;
+
+/** Provides predicates for constructing flow-summary input specifications */
+module SummaryInput {
+  private import semmle.code.csharp.frameworks.system.Collections
+
+  /**
+   * Gets an input specification that specifies the `i`th parameter as
+   * the input (`i = -1` corresponds to the implicit `this` parameter).
+   */
+  SummaryInput parameter(int i) { result = TParameterSummaryInput(i) }
+
+  private predicate isCollectionType(ValueOrRefType t) {
+    t.getABaseType*() instanceof SystemCollectionsIEnumerableInterface and
+    not t instanceof StringType
+  }
+
+  /**
+   * Gets an input specification that specifies the `i`th parameter as
+   * the input (`i = -1` corresponds to the implicit `this` parameter).
+   *
+   * `inputContents` is either empty or a singleton element content list,
+   * depending on whether the type of the `i`th parameter of `c` is a
+   * collection type.
+   */
+  SummaryInput parameter(SummarizableCallable c, int i, ContentList inputContents) {
+    result = parameter(i) and
+    exists(Parameter p |
+      p = c.getParameter(i) and
+      if isCollectionType(p.getType())
+      then inputContents = ContentList::element()
+      else inputContents = ContentList::empty()
+    )
+  }
+
+  /**
+   * Gets an input specification that specifies the implicit `this` parameter
+   * as the input.
+   */
+  SummaryInput thisParameter() { result = TParameterSummaryInput(-1) }
```

Comment:
What are the consequences / advantages of `this` being equivalent to parameter `-1`?  In CPP we have something similar to `thisParameter`, but it returns a value from a different branch of the `newtype` rather than overloading the meaning of a parameter.

Reply:
It aligns with how `this` parameters are modelled in the data-flow library for C#. We are not exposing the IPA type, so I don't think it should matter to the end-user.